### DISCUSSION TITLE
[RFC] strapi-plugin-utils

### DISCUSSION THREAD

**Discussion by ComfortablyCoding** [strapi-community/.github#6] [Ideas]
# Overview

This RFC is aimed at opening a discussion regarding a utility package for strapi plugin developers to work with.  The package will contain common utility functions that plugin developers are consistently in the need of in their own projects.

# Motivation
The motivation behind this is to provide a central location for all utilities that a plugin developer might need.

# Design
The current design will be as follows.

## Language
The langauge for the package will be in typescript to provide the ability to build for both commonjs and esm support in addition to the other benefits of using typescript. This will allow the plugin to provide utilities for both the frontend and backend of a plugin.

## Structure
The plan is to have a plugin-utils package and a test-utils package each published seperately via NPM.

### plugin-utils
The plugin utils package will contain the common utility functions. The structure of the package itself is as follows:
- server - contains all utility functions that can be utlitized on the server side/backend
- admin- contains all utility functions that can be utlitized on the admin side/frontend
- polymorphic/shared/common - contains all utility functions that can be utlitized on the both the admin and the server

### Utility Access
The utilities themselves can be access via the following code
```js
// folder scoped
import {
    response
} from '@strapi-community/plugin-utils/admin';

function fetchArticles = async () => {
    const res = await axios('/api/articles');
    if (response.hasError(data) || response.isEmpty(res)) {
        return [];
    }

    return res.data
}

// variable scoped
import {
    admin
} from '@strapi-community/plugin-utils';

function fetchArticles = async () => {
    const res = await axios('/api/articles');
    if (admin.response.hasError(data) || admin.response.isEmpty(res)) {
        return [];
    }

    return res.data
}
```
### Current Utiltites
- shared
  - field
    - isDynamicZone
    - isManyRelation
    - isMultiMedia
    - isRepeatableComponent
    - isSingleComponent
    - isSingleMedia
    - isSingleRelation
  - response
    - hasError
    - isCollection
    - isEmpty
    - isSingle 
- server 
  - entity
    - isPublished
  - model
    -  hasDraftAndPublishEnabled
### test-utils
This package will contain common utilities needed for plugin tests.

### Primer
Let me know what everyone thinks about this and any feedback regarding structure, additional utilities etc. Interested to get this thing moving!

**Comment by derrickmehaffy** [comment]
This sounds like a fantastic idea :)

**Comment by stafyniaksacha** [comment]
I will also add `app-utils` where we can find utils related to bootsrap / fixtures / cron job management etc...

**Reply by ComfortablyCoding** [reply]
In that case we should probably name the repo something more general, probably `utils` or `strapi-utils` would be more apt.

**Reply by Eventyret** [reply]
I agree with @ComfortablyCoding on this.
Utils should hold or strapi-community-utils incase we make our own package scope etc.
@strapi-community/utils

**Reply by derrickmehaffy** [reply]
The `@strapi-community/utils` would be great honestly

**Reply by derrickmehaffy** [reply]
Falls in with the fact that we too also have a utils package: https://github.com/strapi/strapi/tree/master/packages/core/utils
https://www.npmjs.com/package/@strapi/utils

**Comment by ComfortablyCoding** [comment]
I am thinking we should have 3 plugin util packages instead of just one with multi folders.

- plugin-utils-admin
- plugin-utils-server
- plugin-utils-shared


This will allow for per package configuration as well as reduce unnecessary code.

For example with the current strategy if a user only needs the admin utils they would get the server and shared as well with a single package setup.

Only downside i see here is the need to install an additional two packages if a user wants everything.

Not sure if this is needed for test utils as well. Depends on where we go with that.

Any thoughts?

P.S. I plan to have first draft ready to be pushed this weekend.

**Reply by derrickmehaffy** [reply]
We could build it as a mono-repo and just distribute the packages much like AWS does with their sdk? (which largely functions as a utils sdk)

**Comment by ComfortablyCoding** [comment]
I have uploaded a first draft of the repo on my own account named [strapi-utils](https://github.com/ComfortablyCoding/strapi-utils). We can move it to the `@strapi-community/utils` repo once everything is ironed out. 

Let me know if anything else is needed before it can be migrated to strapi-community. At the moment it is a bit rough and not ready for public consumption (although all tests are passing :smile: )

**Comment by stafyniaksacha** [comment]
We created the https://github.com/strapi-community/utils repository !
